portfolio/src/main/webapp/index.html

```diff
@@ -108,13 +108,23 @@ <h3 id="contact_me">Contact Me</h3>
                placeholder="example@gmail.com"><br>
         <input class="button" type="submit" value="Submit">
       </form>
+      <div id="is-latest-input-dangerous"></div>
       <p><b>Comment History</b></p>
-      <label for="maxCommentsToDisplay">Maximum number of entries to show in comment history:</label>
-      <input type="number" id="maxCommentsToDisplay" name="maxCommentsToDisplay"
-             onchange="addComments()" min="0" max="" value="">
-      <p>(The latest comments are displayed on top.)</p>
       <ul id="comment-container"></ul>
-      <button onclick="addComments()">Refresh</button>
+      <div id="pagination">
+        <label for="max-comments-to-display">Maximum number of entries to show on one page of the comment history:</label>
+        <input type="number" id="max-comments-to-display" name="max-comments-to-display" onchange="addComments(1)"
+               min="0" max="" value="">
+        <p>Currently on page <b id="current-page-id"></b>. (The latest comments are displayed on top.)</p>
+        <button id="first" onclick="addComments(this.id)">&lt;&lt;</button>
+        <button id="prev" onclick="addComments(this.id)">&lt;</button>
+        <button id="next" onclick="addComments(this.id)">&gt;</button>
+        <button id="last" onclick="addComments(this.id)">&gt;&gt;</button>
+        <label for="go-to-page-id">Go to page </label>
+        <input type="number" id="go-to-page-id" name="go-to-page-id" onchange="addComments(this.value)"
```

Comment:
I would also add some validation here around not allowing it to go over the "max number of pages".  For instance, if there is 1 page I can still enter "5" here and have it tell me I'm on page 5 lol

Reply:
HTML's max attribute only sets the biggest number users can get for using the up arrow on the right side of the text box. Users can still type in larger numbers. To do further validation, I would probably need JavaScript or backend Java code. I'll look into this.

Reply:
Currently my code checks for this on the backend Java side. I'll try to debug that code and see if I can improve that behavior. Thanks for pointing this out! I think this is quite important

Reply:
The error checking works now, with a warning message displayed in the following 2 scenarios:
1. user enters a negative maximum number of comments to display;
2. user enters a page ID out of the range [1, total number of pages].

Implementation-wise, because I need to determine the valid range by first querying the backend database for the total number of comments, checking user input must involve the Java servlet. Therefore, I kept the error checking logic in the Java servlet. There were some bugs and missed edge cases with the code (such as with 0 max comments to display, or in situations when there is no comment/page in the database); they are now fixed. Next time, I'll definitely remember to test thoroughly.

The Java servlet is completely in charge of the logic to check user input's validity. It indicates this validity through 2 flags. The client-side code is in charge of checking those flags and showing warning messages if necessary.

Reply:
Ooo great! I wonder if it would make sense to show the max number of pages (i.e. similar to "currently on page 1" you could do something like "page 1 of 10" or whatever)